### ISSUE TITLE
Minify JavaScript files / .min added to paypal file

### ISSUE THREAD

**Issue by exreplay** [magento/magento2#5835]
When "Minify JavaScript Files" is turned on, Magento does add .min to the external paypal script. You can enable merge or bundling, both don't affect the file or in general external files.
### Preconditions
1. Magento 2.1.0
2. PHP-FPM 7.0.8
3. 10.1.16-MariaDB-1~jessie
### Steps to reproduce
1. Login to Backend
2. Shops -> Configuration ->Advanced -> Developer -> JavaScript Settings 
3. Put "Minify JavaScript Files" to "Yes"
4. Save Config
### Expected result
1. "https://www.paypalobjects.com/api/checkout.js" is included
### Actual result
1. "https://www.paypalobjects.com/api/checkout.min.js" is included

**Comment by kandy** [comment]
Thanks for reporting, but https://www.paypalobjects.com/api/checkout.min.js return 404. Looks like expected result are wrong so I am closing this issue, as fixed. Feel free to reopen issue if we missed anything.

**Comment by hostep** [comment]
@kandy: the expected result is without the `.min` in the filename, I think you switched expected with actual results?

Maybe best to reopen and confirm if this is an actual problem (it does sounds like a possible bug).

**Comment by kandy** [comment]
oh, sorry, my miss

**Comment by exreplay** [comment]
No problem.
The main problem is, when PayPal is integrated and you turn on the minification, sites break because of that.
![paypal_minify_bug](https://cloud.githubusercontent.com/assets/4998475/17215425/15a7a66e-54dd-11e6-9dda-5872a2f5e139.jpg)

**Comment by Sam-PUMP** [comment]
Having the same issue.

**Comment by mengento** [comment]
Temporary solution
Change line 243 of Magento/Framework/RequireJs/Config.php

`Url = url.replace (/ (\. Min)? \. Js $ /, '.min.js');`
To:
`If (url == 'https://www.paypalobjects.com/api/checkout.js') {`
`Url = url;`
`} Else {`
`Url = url.replace (/ (\. Min)? \. Js $ /, '.min.js');`
`}}`

**Comment by AirmanAJK** [comment]
Magento CE 2.1.2 here. Is this getting fixed any time soon? Do I really have to choose between hacking a core file, not using minified javascript, or not using PayPal?

**Comment by AirmanAJK** [comment]
I have a more appropriate solution. Magento 2 has a "minify exclusion list" that can be stored in the core_config_data table.

First make sure there isn't already an entry from some other extension:

select *
from core_config_data ccd
where ccd.path = 'dev/js/minify_exclude';

If there is, you need to append your rule preceded by a line break character (that's what the code splits on).

If it's empty, insert it with a value excluding the paypal checkout file:

insert into core_config_data(scope, scope_id, path, value)
values ('default', 0, 'dev/js/minify_exclude', 'api/checkout');

I believe the proper solution for the Magento devs is to add the following to the Paypal module's config.xml:

```
<dev>
    <js>
        <minify_exclude>
            api/checkout
        </minify_exclude>
    </js>
</dev>
```

**Comment by sshymko** [comment]
Experiencing the same issue.

Ideally, only URLs that belong to the store (start with the store's BaseURL) need to be processed.
A quick workaround in `\Magento\Framework\RequireJs\Config::getMinResolverCode()`:
```js
if (url.indexOf(window.location.hostname) && {$excludesCode}) {
    url = url.replace(/(\.min)?\.js$/, '.min.js');
}
```

**Comment by ghost** [comment]
Is it possible to extend or apply plugins to classes in magento/framework?

**Comment by iivashchenko** [comment]
Hi @exreplay, thank you for reporting. I've created internal ticket MAGETWO-62713 and linked this GitHub issue.

**Comment by rhoerr** [comment]
We're experiencing the same issue for a third-party library a module includes.

**Comment by Igloczek** [comment]
Response from M support team, delivered by @alankent :

They need to add configuration with specific regular expressions to extension's config.xml:
```xml
<?xml version="1.0"?>
<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:module:Magento_Store:etc/config.xsd">
    <default>
        <dev>
            <js>
                <minify_exclude>
                    someexternalfileregex
                    /tiny_mce/
                </minify_exclude>
            </js>
        </dev>
    </default>
</config>
```

Magento doesn't collect and implode all values for xpath `dev/js/minify_exclude` from `config.xml` file of all extensions.
Value for `dev/js/minify_exclude` should be set once for all files which should be excluded.
So merchant create separate extension for setting this configuration in one place.
Also it may be set in `core_config_data` table as row like:
```sql
INSERT INTO `core_config_data` (`scope`, `scope_id`, `path`, `value`)
VALUES ('default', 0, 'dev/js/minify_exclude', ' external\n /tiny_mce/');
```

**Comment by omiroshnichenko** [comment]
Looks like Paypal have been add minified version of file, so issue cannot be reproduced.

**Comment by rhoerr** [comment]
It may not break PayPal anymore, but this issue still affects third party libraries.

Addendum to @Igloczek 's note above with the workaround: If your library is included by requirejs (IE, not a header script tag), the filename will be processed via the following script (generated by PHP at runtime):

`/pub/static/adminhtml/Magento/backend/en_US/secure/requirejs-min-resolver.min.js`

The `dev/js/minify_exclude` setting is baked into that file, meaning any change to the setting requires purging and redeploying static content to actually take effect.

**Comment by martinabenedetto** [comment]
Hi, Magento 2.3 here. It seems the <minify_exlude> sintax has changed now. 
It looks like:
`                <minify_exclude>
                    <tiny_mce>/tiny_mce/</tiny_mce>
                </minify_exclude>`

The point is that adding a new line to exclude external js files to (e.g. Vimeo api) to be minified is not working anymore:
`<vimeo>player.vimeo.com/api/player</vimeo>`

Any idea?

**Comment by hostep** [comment]
@MartinaB90: that's correct. The syntax to exclude certain url's from getting the `.min` added in the filename changed in 2.3.0 by https://github.com/magento/magento2/pull/13687
That was to make sure multiple modules could add their own exclusion without overwriting other modules configurations.

At first look, your Vimeo regular expression seems to be correct.
Can you give a bit of context of how you include the Vimeo player javascript file in the page, so I can try to reproduce the problem?

Thanks!

**Comment by AndyJAllen** [comment]
> 
> 
> Temporary solution
> Change line 243 of Magento/Framework/RequireJs/Config.php
> 
> `Url = url.replace (/ (\. Min)? \. Js $ /, '.min.js');`
> To:
> `If (url == 'https://www.paypalobjects.com/api/checkout.js') {`
> `Url = url;`
> `} Else {`
> `Url = url.replace (/ (\. Min)? \. Js $ /, '.min.js');`
> `}}`

is there any reason why we shouldn't do this for all external url addresses? Just see if domain name <> your websites domain.

It seems like there is a lot of headache and hassle over solving an issue when magento shouldn't even be trying to minify off-site links to javascript files.

**Comment by martinabenedetto** [comment]
Hi guys, I solved the issue (in 2.3) excluding the full web address of Vimeo external resource. It is working now. I think this is the correct solution since you'll be able to manage the exclusion in your module and not changing Config.php.

`    <default>
        <dev>
            <js>
                <minify_exclude>
                    <vimeo>https://player.vimeo.com/api/player.js</vimeo>
                </minify_exclude>
            </js>
        </dev>
    </default>`

**Comment by wltprgm** [comment]
I literally have no idea what kind of configuration you guys are talking about. And nobody seems to explain what this kind of syntax does.

<tiny_mce>/regex/</tiny_mce>
<vimeo>https://someurl/player.js</vimeo>

When to put certain type of tags?

I am trying to modify an extension from Mageplaza because of this issue during minification.

https://www.mageplaza.com/magento-2-social-share/

![image](https://user-images.githubusercontent.com/27481268/61994323-04a07980-b0ac-11e9-87f5-dbde27200e85.png)


adding
```
        <dev>
            <js>
                <minify_exclude>
                    <config>https://static.addtoany.com/menu/page.js</config>
                </minify_exclude>
            </js>
        </dev>
```
to the module config.xml don't work at all
